### PR TITLE
gitlab_project_variable: fix nondeterministic reads when using environment_scope

### LINKED ISSUE #195: [Bug] environment_scope for resource gitlab_project_variable defaults to 0 instead of *
Hi,

We have run into an issue with the gitlab_project_variable resource. We have several Gitlab variables defined using Terraform. After the upgrade to 0.12.11 the variables were not set in the CI pipeline. After investigating we found that the new version automatically set the environment_scope to "0" which results in a new scope listed in the Gitlab CI/CD variable settings. This new scope is not used by the CI/CD pipeline resulting in missing variable values during the deployment. After adding `environment_scope = "*"` to the Terraform script the correct scope "All environments" was set.

Kind regards,
Kristian

### Terraform Version

Docker image hashicorp/terraform with version 0.12.11
Plugin for provider "gitlab" (terraform-providers/gitlab) 2.3.0

### Affected Resource(s)

- environment_scope

### Terraform Configuration Files
```hcl
resource "gitlab_project_variable" "api_set_logging_s3_bucketname_staging" {
  key     = "AWS_S3_BUCKET_NAME_STAGING"
  project = gitlab_project.api.id
  value   = aws_s3_bucket.s3_logging_staging.bucket
}
```

### Debug Output
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

# gitlab_project_variable.api_set_logging_s3_bucketname_staging will be updated in-place
  ~ resource "gitlab_project_variable" "api_set_logging_s3_bucketname_staging" {
      + environment_scope = "0"
        id                = "90:AWS_S3_BUCKET_NAME_STAGING"
        key               = "AWS_S3_BUCKET_NAME_STAGING"
        masked            = false
        project           = "90"
        protected         = false
        value             = (sensitive value)
        variable_type     = "env_var"
    }
``` 

### Expected Behavior

* If not defined in resource the environment_scope value should be set to "*" for All environments

### Actual Behavior

* environment_scope is set to non-existing scope "0"

### Steps to Reproduce

1. Create gitlab_project_variable resource with key, project and value (was created with older version in Gitlab by older Terraform version )
2. Run terraform plan

### LINKED ISSUE #213: Incorrect project environment variable lookup
### Terraform Version
```
$ terraform -v
Terraform v0.12.13
+ provider.vault v2.6.0
```

### Affected Resource(s)
Please list the resources as a list, for example:
- gitlab_project_variable

### Terraform Configuration Files
```hcl
resource "gitlab_project_variable" "vault-role-id" {
  key               = "VAULT_ROLE_ID"
  value             = vault_approle_auth_backend_role.app-role.role_id
  project           = var.project.gitlab_project.id
  environment_scope = "development"
  protected         = var.gitlab_protected_environment
}

resource "gitlab_project_variable" "vault-role-id" {
  key               = "VAULT_ROLE_ID"
  value             = vault_approle_auth_backend_role.app-role.role_id
  project           = var.project.gitlab_project.id
  environment_scope = "production"
  protected         = var.gitlab_protected_environment
}
```

### Description
The GitLab project environment variable lookup is incorrect. In GitLab an environment variable can have the same name in multiple environment scopes. However, the GitLab api doesn't support this on the REST api that is used in the source code of this provider. So what happens is that the wrong environment variable is being read when refreshing the state.

Solving this would be to fetch the list of all project variable and find the variable with the corresponding name and environment_scope.
```
/api/v4/projects/<project_id>/variables
```

### LINKED ISSUE #278: Recreate gitlab_project_variable if it was manually deleted
Hello,

I just ran into the issue that planning fails with the following error if a previously created project variable has been manually deleted. It would be better if the provider recreated the variable instead of failing.

```
Error: GET https://gitlab.example.com/api/v4/projects/81/variables/MY_VARIABLE: 404 {message: 404 Variable Not Found}
```

- Terraform version: 0.12.23
- GitLab provider version: 2.5.0
- Affected Resource(s)
  - gitlab_project_variable
  - gitlab_group_variable ~~(likely)~~ (Edit: Confirmed. See first comment)

### PR DESCRIPTION
**Pull request by armsnyder**
Supersedes  #324
Fixes #213
Fixes #195
Fixes #278

**tl;dr** I borrowed just the read function logic from #324 and added the `environment_scope` filter query to the update and delete functions.

This PR fixes a behavior where `gitlab_project_variable` resources for keys with multiple environment scopes were not reading data from a unique variable, but could read from any variable with the specified key, causing unexpected drift.

For users with the [feature flag on](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/34490) or who are using [GitLab 13.4+](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/39209), this also lets them change environment scopes and remove variables with environment scopes, for `gitlab_project_variable` resources that have non-unique keys.

Note that one big difference between this and #324 is that I decided to only modify the read function, and not modify the update or delete functions, apart from adding the `environment_scope` filter query. I thought that those functions as implemented in #324 were not idempotent-safe. I would rather wait for a proper fix from GitLab (see previous paragraph) than try to work around the API in the provider for older GitLab versions. The `environment_scope` filter query in those functions ensures that people using the feature flag or on future versions of GitLab will have a better time.

I tried using a `MigrateState` in the schema to handle the ID format change, but it didn't work when I tested it. I ended up with an ID format error. Instead, I have the read function handle both ID formats.

Note that the `gitlab/resource_gitlab_project_variable_test.go` file is easier to review side-by-side than inline, since I rewrote it from scratch to make sure it was exercising all behaviors.

I've tested this locally with GitLab v13.1 and v13.4 to make sure both versions work.

Incidentally this PR is a step toward #205, by having a destroy test that leaves the GitLab project alone and checks that the resources under test are removed from GitLab.

### CODE REVIEW

**Comment by armsnyder** [comment]
@roidelapluie @nicholasklick @sfang97 This is the continuation of #324, which has merge conflicts. I simplified the resource implementation greatly from that PR (see PR description for why). On the other hand the test file is a lot to review, but I wanted to make sure the resource was fully covered. ❤️

File: website/docs/r/project_variable.html.markdown [hunk 1]
```diff
@@ -12,6 +12,7 @@ This resource allows you to create and manage CI/CD variables for your GitLab pr
 For further information on variables, consult the [gitlab
 documentation](https://docs.gitlab.com/ce/ci/variables/README.html#variables).
 
+~> **Important:** If your GitLab version is older than 13.4, you may see nondeterministic behavior when updating or deleting `gitlab_project_variable` resources with non-unique keys, for example if there is another variable with the same key and different environment scope.
```
Comment (roidelapluie) [inline website/docs/r/project_variable.html.markdown:15]: Can you link to relevant issues?

File: gitlab/util.go [hunk 2]
```diff
@@ -181,11 +182,26 @@ func parseTwoPartID(id string) (string, string, error) {
...
+	return parts[0], parts[1], parts[2], nil
+}
+
 // format the strings into an id `a:b`
 func buildTwoPartID(a, b *string) string {
 	return fmt.Sprintf("%s:%s", *a, *b)
 }
 
+// buildMultiPartID formats the strings into an id `a:b:c`, taking any number of parts.
+func buildMultiPartID(parts ...string) string {
```
Comment (roidelapluie) [inline gitlab/util.go:201]: That seems an un-necessary indirection

File: website/docs/r/project_variable.html.markdown (hunk 1 above)
Comment (armsnyder) [reply website/docs/r/project_variable.html.markdown]: Done

File: gitlab/util.go (hunk 2 above)
Comment (armsnyder) [reply gitlab/util.go]: Removed

**Comment by roidelapluie** [comment]
awesome

**Review by roidelapluie — APPROVED** [review]

**Comment by nicholasklick** [comment]
@roidelapluie @armsnyder if this is good to go and we merge it?

**Comment by abelmokadem** [comment]
Any ETA for the release?

**Comment by armsnyder** [comment]
@abelmokadem Soon™! It's our first time releasing to the Terraform Registry, but we're making progress in the #411 issue.

File: website/docs/r/project_variable.html.markdown [hunk 3]
```diff
@@ -12,6 +12,10 @@ This resource allows you to create and manage CI/CD variables for your GitLab pr
 For further information on variables, consult the [gitlab
 documentation](https://docs.gitlab.com/ce/ci/variables/README.html#variables).
 
+~> **Important:** If your GitLab version is older than 13.4, you may see nondeterministic behavior
```
Comment (abelmokadem) [inline website/docs/r/project_variable.html.markdown:15]: 13.2.0 you mean?
